## QXD0178 - Mineração de Dados
# Classificação de dados

# Lista de Exercícios: Classificação de dados

Nesta lista de exercícios, você explorará a aplicação de métodos de aprendizado de máquina para realizar tarefas de classificação de dados. Você usará a base de dados [Food choices: College students' food and cooking preferences](https://www.kaggle.com/datasets/borapajo/food-choices?select=food_coded.csv) e avaliará vários algoritmos de classificação para determinar sua eficácia. O objetivo é entender como diferentes métodos de aprendizado de máquina se comportam em relação à acurácia na classificação de dados.

O exercício será dividido em várias etapas:

1. **Pré-processamento dos dados:**
   - Descreva brevemente o conjunto de dados   
   - Limpe o conjunto de dados, tratando valores ausentes, removendo duplicatas e realizando transformações necessárias. 
   - Caso você use os dados pré-processados na lista anterior, faça um breve descritivo dos principais ajustes.
   - Codifique variáveis categóricas, se necessário, para que possam ser utilizadas em algoritmos de aprendizado de máquina.
   - Cria a coluna `self_perception_overweight` com valor: `True` se a coluna `self_perception_weight` tem valor 4 ou 5; e `False`, caso contrário.
   - Remova a coluna `self_perception_weight` do conjunto de dados.
2. **Divisão do conjunto de dados:**
   - Divida o conjunto de dados em um conjunto de treinamento e um conjunto de teste para avaliar o desempenho dos algoritmos. 
   - O mesmo conjunto de teste deve ser usado por todos os algoritmos analizados e nenhum dado deste pode ser usado na fase de treinamento.
   - O atributo alvo (*rótulo*) da classificação será o campo `self_perception_overweight`.   
3. **Seleção de algoritmos de classificação:**
   - Selecione uma variedade de algoritmos de aprendizado de máquina para testar na tarefa de classificação.   
   - Sua seleção deve conter, no mínimo, os seguintes métodos: Naive Bayes, k-Nearest Neighbors, Support Vector Machine (Linear/RBF), Decision Trees, Random Forest, Multilayer Perceptron.
   - Descreva brevemente como funciona cada algoritmo selecionado.
4. **Treinamento e avaliação:**
   - Treine os algoritmos de classificação usando todo o conjunto de treinamento. 
   - Avalie o desempenho de cada algoritmo no conjunto de teste usando métricas como acurácia, precisão, recall e F1-score.
   - Repita a análise treinando os algoritmos com validação cruzada.
   - Repita a análise realizando ajuste de hiperparâmetros.
5. **Análise dos resultados:**
   - Prepare um texto que descreva os resultados obtidos e faça uma análise crítica destes resultados.
   - Compare o desempenho dos diferentes algoritmos e explique por que alguns apresentaram resultados mais adequados que outros.
   
Documente todas as etapas em um arquivo Jupyter Notebook (`.ipynb`) que inclua as análises, o código e as justificativas. Lembre-se de que é fundamental justificar todas as decisões tomadas ao longo do processo e documentar as análises de forma clara e concisa. Este trabalho tem como objetivo proporcionar uma compreensão prática da seleção e avaliação de algoritmos de classificação em cenários de aprendizado supervisionado.

Envie seu Jupyter Notebook até a data de entrega especificada nesta tarefa.

## Solução


### Pré-processamento dos dados

A base de dados é um conjunto de dados que coleta informações sobre as escolhas alimentares e preferências culinárias de estudantes universitários, ela tem informações desde dados demográficos como sexo e renda, passando por dados relacionados a preferências alimentares como comida favorita, estimativa de quantas calorias possui determinado elemento e etc. As colunas presentes na base de dados são: *'GPA', 'Gender', 'breakfast', 'calories_chicken', 'calories_day','calories_scone', 'coffee', 'comfort_food', 'comfort_food_reasons', 'comfort_food_reasons_coded', 'cook', 'comfort_food_reasons_coded.1', 'cuisine', 'diet_current', 'diet_current_coded', 'drink', 'eating_changes', 'eating_changes_coded', 'eating_changes_coded1', 'eating_out', 'employment', 'ethnic_food', 'exercise', 'father_education', 'father_profession', 'fav_cuisine', 'fav_cuisine_coded', 'fav_food', 'food_childhood', 'fries', 'fruit_day', 'grade_level', 'greek_food', 'healthy_feeling', 'healthy_meal', 'ideal_diet', 'ideal_diet_coded', 'income', 'indian_food', 'italian_food', 'life_rewarding', 'marital_status', 'meals_dinner_friend', 'mother_education', 'mother_profession', 'nutritional_check', 'on_off_campus', 'parents_cook', 'pay_meal_out', 'persian_food', 'self_perception_weight', 'soup', 'sports', 'thai_food', 'tortilla_calories', 'turkey_calories', 'type_sports', 'veggies_day', 'vitamins', 'waffle_calories', 'weight'*

Na parte de **Limpeza dos Dados**, parecido com oque foi feito na primeira limpeza do dataset utilizando apenas as bibliotecas numpy e pandas, foi realizada apenas a adapatação de algumas operações para que fossem executadas em uma única célula, a substituição de valores NaN e alguns que não faziam sentido no contexto da coluna localizada na base de dados para valores que fizessem sentido e não gerasse resultados muito tendenciosos. No caso de valores númericos, foram utilizadas estratégias como, por exemplo, a substituição por valores de média e moda. Em valores de string, foram substituídos NaN por uma única constante 'none', considerada como valor vazio. Além disso, para outras strings que não faziam sentido dentro de seu contexto como 'personal' onde esperava-se um resultado, usar outra unidade de peso etc. foi implementado mais uma vez o valor none. Por último, também houve a exclusão de linhas inteiras cujo a maioria dos valores ou até todos eles eram NaN.

In [1]:
import numpy as np 
import pandas as pd

df = pd.read_csv("https://raw.githubusercontent.com/She-Codes-Now/Intro-to-Data-Science-with-R/master/food_coded.csv")
# Preenche valores ausentes na coluna 'soup' com a moda
df['soup'].fillna(df['soup'].mode()[0], inplace=True)

# Preenche valores ausentes na coluna 'sports' com a moda
df['sports'].fillna(df['sports'].mode()[0], inplace=True)

# Calcula a média da coluna 'calories_day' e preenche valores ausentes com a média arredondada
media = df['calories_day'].mean()
media_rounded = round(media, 1)
df['calories_day'].fillna(media_rounded, inplace=True)

# Preenche valores ausentes nas colunas 'tortilla_calories' e 'calories_scone' com a moda
df['tortilla_calories'].fillna(df['tortilla_calories'].mode()[0], inplace=True)
df['calories_scone'].fillna(df['calories_scone'].mode()[0], inplace=True)

# Remove a coluna 'comfort_food_reasons_coded'
del df['comfort_food_reasons_coded']

# Renomeia a coluna 'comfort_food_reasons_coded.1' para 'comfort_food_reasons_coded'
df.rename(columns={'comfort_food_reasons_coded.1': 'comfort_food_reasons_coded'}, inplace=True)

# Calcula a média da coluna 'cook' e preenche valores ausentes com a média arredondada
media = df['cook'].mean()
media_rounded = round(media, 0)
df['cook'].fillna(media_rounded, inplace=True)

# Preenche valores ausentes na coluna 'cuisine' com a moda
df['cuisine'].fillna(df['cuisine'].mode()[0], inplace=True)

# Calcula a média da coluna 'drink' e preenche valores ausentes com a média arredondada
media = df['drink'].mean()
media_rounded = round(media, 0)
df['drink'].fillna(media_rounded, inplace=True)

# Calcula a média da coluna 'employment' e preenche valores ausentes com a média arredondada
media = df['employment'].mean()
media_rounded = round(media, 0)
df['employment'].fillna(media_rounded, inplace=True)

# Calcula a média da coluna 'exercise' e preenche valores ausentes com a média arredondada
media = df['exercise'].mean()
media_rounded = round(media, 0)
df['exercise'].fillna(media_rounded, inplace=True)

# Calcula a média da coluna 'father_education' e preenche valores ausentes com a média arredondada
media = df['father_education'].mean()
media_rounded = round(media, 0)
df['father_education'].fillna(media_rounded, inplace=True)

# Preenche valores ausentes na coluna 'fav_food' com a moda
df['fav_food'].fillna(df['fav_food'].mode()[0], inplace=True)

# Calcula a média da coluna 'income' e preenche valores ausentes com a média arredondada
media = df['income'].mean()
media_rounded = round(media, 0)
df['income'].fillna(media_rounded, inplace=True)

# Calcula a média da coluna 'life_rewarding' e preenche valores ausentes com a média arredondada
media = df['life_rewarding'].mean()
media_rounded = round(media, 0)
df['life_rewarding'].fillna(media_rounded, inplace=True)

# Preenche valores ausentes na coluna 'marital_status' com a moda
mode = df['marital_status'].mode()[0]
df['marital_status'].fillna(mode, inplace=True)
df.at[74, 'marital_status'] = mode

# Calcula a média da coluna 'mother_education' e preenche valores ausentes com a média arredondada
media = df['mother_education'].mean()
media_rounded = round(media, 0)
df['mother_education'].fillna(media_rounded, inplace=True)

# Preenche valores ausentes na coluna 'on_off_campus' com a moda
df['on_off_campus'].fillna(df['on_off_campus'].mode()[0], inplace=True)

# Calcula a média da coluna 'persian_food' e preenche valores ausentes com a média arredondada
media = df['persian_food'].mean()
media_rounded = round(media, 0)
df['persian_food'].fillna(media_rounded, inplace=True)

# Calcula a média da coluna 'self_perception_weight' e preenche valores ausentes com a média arredondada
media = df['self_perception_weight'].mean()
media_rounded = round(media, 0)
df['self_perception_weight'].fillna(media_rounded, inplace=True)

# Define o valor da linha 2 na coluna 'weight' como NaN
df.at[2, 'weight'] = np.nan

# Extrai valores numéricos da coluna 'weight', converte para float e preenche valores ausentes com a média arredondada
df['weight'] = df['weight'].str.extract('(\d+)').astype(float)
media = df['weight'].mean()
media_rounded = round(media, 0)
df.at[2, 'weight'] = media_rounded
df['weight'].fillna(media_rounded, inplace=True)

# Define valores NaN nas linhas 61 e 104 da coluna 'GPA'
df.at[61, 'GPA'] = np.nan
df.at[104, 'GPA'] = np.nan

# Extrai valores numéricos da coluna 'GPA', converte para float e preenche valores ausentes com a média
df['GPA'] = df['GPA'].str.extract('(\d+)').astype(float)
media = df['GPA'].mean()
df['GPA'].fillna(media, inplace=True)

# Remove a linha 74 do DataFrame e redefine os índices
df = df.drop(74).reset_index(drop=True)

# Preenche valores ausentes em várias colunas categóricas com a string 'none'
df['eating_changes'].fillna('none', inplace=True)
df['father_profession'].fillna('none', inplace=True)
df['meals_dinner_friend'].fillna('none', inplace=True)
df['mother_profession'].fillna('none', inplace=True)
df['type_sports'].fillna('none', inplace=True)

<>:94: SyntaxWarning: invalid escape sequence '\d'
<>:105: SyntaxWarning: invalid escape sequence '\d'
<>:94: SyntaxWarning: invalid escape sequence '\d'
<>:105: SyntaxWarning: invalid escape sequence '\d'
C:\Users\elyss\AppData\Local\Temp\ipykernel_18088\2104592061.py:94: SyntaxWarning: invalid escape sequence '\d'
  df['weight'] = df['weight'].str.extract('(\d+)').astype(float)
C:\Users\elyss\AppData\Local\Temp\ipykernel_18088\2104592061.py:105: SyntaxWarning: invalid escape sequence '\d'
  df['GPA'] = df['GPA'].str.extract('(\d+)').astype(float)
C:\Users\elyss\AppData\Local\Temp\ipykernel_18088\2104592061.py:6: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.metho

#### Normalização

In [2]:
# Colocando todas as strings em minúsculas
df = df.applymap(lambda x: x.lower() if isinstance(x, str) else x)
# Tirando os espaços em branco
df = df.applymap(lambda x: x.replace(" ", "") if isinstance(x, str) else x)

C:\Users\elyss\AppData\Local\Temp\ipykernel_18088\343034208.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  df = df.applymap(lambda x: x.lower() if isinstance(x, str) else x)
C:\Users\elyss\AppData\Local\Temp\ipykernel_18088\343034208.py:4: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  df = df.applymap(lambda x: x.replace(" ", "") if isinstance(x, str) else x)


foi criada uma tabela que representa se o estudante pratica ou não esportes. Isso ajudará o algoritmo pois há a tendência natural que praticar esportes tende à dificultar a existência de sobrepeso.

In [3]:
df['practice_sports'] = np.where(df['type_sports'] != 'none', 1, 0)
df['practice_sports']

0      1
1      1
2      0
3      0
4      1
      ..
119    1
120    1
121    0
122    0
123    0
Name: practice_sports, Length: 124, dtype: int64

"criar categorias númericas para os tipos de culinária favoritos dos estudantes?". Pode haver uma tendência de determinada culinária ocasionar uma sensação de sobrepeso. Então a biblioteca LabelEncoder transforma valores categóricos em valores numéricos, mapeando cada valor categórico único em 'fav_cuisine' para um número inteiro, começando de 0 até n_classes-1, sendo o n a quantidade de categorias únicas, e assim armazenado em uma nova coluna que nela contém os valores numéricos correspondentes às categorias originais, apenas exibindo as colunas com valores transformados

In [4]:
from sklearn.preprocessing import LabelEncoder

le = LabelEncoder()
df['fav_cuisine_cat'] = le.fit_transform(df['fav_cuisine'])
df['fav_cuisine_cat']

0       6
1      23
2      23
3      45
4      23
       ..
119    23
120    36
121    30
122    23
123    17
Name: fav_cuisine_cat, Length: 124, dtype: int64

As comidas que mais aparecem na tabela *comfort_food*, a quantidade 10 oi um valor considerável para o tamanho da base de dados, então, criando-se uma nova coluna numérica que verifica se aquele estudante tem como comida "confortável" plo menos 1 dessas comidas, se for confirmado que o estudante tem uma comida favorita, com tendência, tem chances de ser acima do peso, sendo um bom parâmetro para o modelo.

In [5]:
from collections import Counter

# Concatena todos os valores da coluna 'comfort_food' em uma única string e divide em uma lista de alimentos
foods = ','.join(df['comfort_food']).split(',')

# Conta a frequência de cada alimento na lista
couting_foods = Counter(foods)

# Cria uma lista de alimentos que aparecem 10 ou mais vezes
foods_more_than_10 = [food for food, counting in couting_foods.items() if counting >= 10]

# Exibe a lista de alimentos populares
foods_more_than_10

# Define uma função para verificar se um registro contém alimentos populares
def find_food(s):
    foods = s.split(',')
    for food in foods:
        if food in foods_more_than_10:
            return 1
    return 0

# Aplica a função 'find_food' a cada registro da coluna 'comfort_food' e cria uma nova coluna 'got_popular_comfort_food'
df['got_popular_comfort_food'] = df['comfort_food'].apply(find_food)

# Exibe a nova coluna com os valores 1 ou 0
df['got_popular_comfort_food']

0      0
1      1
2      1
3      1
4      1
      ..
119    1
120    0
121    0
122    1
123    1
Name: got_popular_comfort_food, Length: 124, dtype: int64

mesmo processo para a *food_childhood*, a infância costuma influênciar na tendência de saúde

In [6]:
from collections import Counter
foods = ','.join(df['food_childhood']).split(',')
couting_foods = Counter(foods)
foods_more_than_10 = [food for food, counting in couting_foods.items() if counting >= 10]
foods_more_than_10

def find_food(s):
    foods = s.split(',')
    for food in foods:
        if food in foods_more_than_10:
            return 1
    return 0

df['got_popular_food_childhood'] = df['food_childhood'].apply(find_food)
df['got_popular_food_childhood']

0      0
1      0
2      1
3      1
4      1
      ..
119    0
120    0
121    0
122    0
123    0
Name: got_popular_food_childhood, Length: 124, dtype: int64

#### Remover os dados aberto

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 124 entries, 0 to 123
Data columns (total 64 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   GPA                         124 non-null    float64
 1   Gender                      124 non-null    int64  
 2   breakfast                   124 non-null    int64  
 3   calories_chicken            124 non-null    int64  
 4   calories_day                124 non-null    float64
 5   calories_scone              124 non-null    float64
 6   coffee                      124 non-null    int64  
 7   comfort_food                124 non-null    object 
 8   comfort_food_reasons        123 non-null    object 
 9   cook                        124 non-null    float64
 10  comfort_food_reasons_coded  124 non-null    int64  
 11  cuisine                     124 non-null    float64
 12  diet_current                124 non-null    object 
 13  diet_current_coded          124 non

In [8]:
df.drop('comfort_food', axis=1, inplace=True)
df.drop('comfort_food_reasons', axis=1, inplace=True)
df.drop('diet_current', axis=1, inplace=True)
df.drop('eating_changes', axis=1, inplace=True)
df.drop('father_profession', axis=1, inplace=True)
df.drop('fav_cuisine', axis=1, inplace=True)
df.drop('food_childhood', axis=1, inplace=True)
df.drop('healthy_meal', axis=1, inplace=True)
df.drop('ideal_diet', axis=1, inplace=True)
df.drop('meals_dinner_friend', axis=1, inplace=True)
df.drop('mother_profession', axis=1, inplace=True)
df.drop('type_sports', axis=1, inplace=True)

criar uma colina *self_perception_overweight* para remover a *self_perception_weight*, percepção acima do peso e a do peso normal, para se ter uma abordagem preditiva e prever comportamentos alimentares ou riscos de doenças por exemplo.

In [9]:
# Cria uma nova coluna 'self_perception_overweight' que é True se 'self_perception_weight' for 4.0 ou 5.0, indicando percepção de sobrepeso
df['self_perception_overweight'] = (df['self_perception_weight'] == 4.0) | (df['self_perception_weight'] == 5.0)

# Remove a coluna original 'self_perception_weight' do DataFrame
df.drop('self_perception_weight', axis=1, inplace=True)

### Divisão do conjunto de dados

Aqui foi dividido o dataset final para o treinamento dos modelos

In [10]:
from sklearn.model_selection import train_test_split

# Define a variável alvo 'y' como a coluna 'self_perception_overweight'
y = df['self_perception_overweight']

# Define as variáveis preditoras 'X' removendo a coluna 'self_perception_overweight'
X = df.drop(columns=['self_perception_overweight'])

# Divide os dados em conjuntos de treino e teste, com 40% dos dados para teste e uma semente aleatória para reprodutibilidade
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.4, random_state=1)

verifico se o *split* foi feito corretamente

In [11]:
# X_train: Conjunto de dados de treino para as variáveis preditoras
# X_test: Conjunto de dados de teste para as variáveis preditoras
# y_train: Conjunto de dados de treino para a variável alvo
# y_test: Conjunto de dados de teste para a variável alvo
print("Dimensões de X_train:", X_train.shape)
print("Dimensões de X_test:", X_test.shape)
print("Dimensões de y_train:", y_train.shape)
print("Dimensões de y_test:", y_test.shape)

print("\nInformações de X_train:")
print(X_train.info())

print("\nInformações de X_test:")
print(X_test.info())

print("\nInformações de y_train:")
print(y_train.info())

print("\nInformações de y_test:")
print(y_test.info())

Dimensões de X_train: (74, 51)
Dimensões de X_test: (50, 51)
Dimensões de y_train: (74,)
Dimensões de y_test: (50,)

Informações de X_train:
<class 'pandas.core.frame.DataFrame'>
Index: 74 entries, 91 to 37
Data columns (total 51 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   GPA                         74 non-null     float64
 1   Gender                      74 non-null     int64  
 2   breakfast                   74 non-null     int64  
 3   calories_chicken            74 non-null     int64  
 4   calories_day                74 non-null     float64
 5   calories_scone              74 non-null     float64
 6   coffee                      74 non-null     int64  
 7   cook                        74 non-null     float64
 8   comfort_food_reasons_coded  74 non-null     int64  
 9   cuisine                     74 non-null     float64
 10  diet_current_coded          74 non-null     int64  
 11  drink         

### Seleção de algoritmos de classificação


**Naive Bayes**
Algoritmo probabilístico que se baseia no teorema de Bayes para realizar tarefas de classificação. Ele assume que as caracteristicas de um determinado dado são independentes umas das outras. É rotineiramente usado para classificar dados de diversos tipos, como numéricos, categóricos e mistos.

**K-Nearest Neighbors**
Algoritmo que classifica uma nova observação com base nas observações mais próximas dela no espaço de atributos. Esse número de observações sendo especificado pelo hiperparâmetro K.

**Support Vector Machine**
Algoritmo que consiste em encontrar um hiperplano de separação ótimo entre duas classes em um espaço de atributos. É amplamente utilizado em tarefas de classificação binária, onde o objetivo é separar as observações em duas classes distintas.

**Decision Tree**
Algoritmo que classifica dados criando uma árvore de decisão. A árvore é construída dividindo o conjunto de dados em subconjuntos cada vez menores, com base nos valores das características.

**Random Forest**
Algoritmo que combina várias árvores de decisão. Isso ajuda o modelo a generalizar melhor para dados novos, reduzindo o risco de sobreajuste

**Multilayer Perceptron**
Algoritmo que usa uma rede neural artificial para classificar dados. Essa rede neural é composta por várias camadas de neurônios, que são conectados uns aos outros por pesos. A rede neural aprende a classificar dados ajustando os pesos dos neurônios.

**Regressão Logística**
Algoritmo que classifica dados usando uma função logística. Tal função é uma função matemática que mapeia um intervalo de números reais para o intervalo ente 0 e 1.Para classificar uma nova observação, o algoritmo de regressão logística calcula a probabilidade de a observação pertencer a cada classe. A observação é então classificada na classe com a probabilidade mais alta.

### Treinamento e avaliação

**Naive Bayes**

In [12]:
from sklearn.naive_bayes import MultinomialNB
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

# Cria um modelo de classificação Naive Bayes Multinomial
model = MultinomialNB()
model.fit(X_train, y_train)

y_pred = model.predict(X_test)

accuracy = accuracy_score(y_test, y_pred)

precision = precision_score(y_test, y_pred)

recall = recall_score(y_test, y_pred)

f1 = f1_score(y_test, y_pred)

print('Accuracy:', accuracy)
print('Precision:', precision)
print('Recall:', recall)
print('F1 Score:', f1)

Accuracy: 0.48
Precision: 0.35294117647058826
Recall: 0.75
F1 Score: 0.48


**KNN**

In [13]:
import os
os.environ['LOKY_MAX_CPU_COUNT'] = '4'
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

# Cria um modelo KNN com 5 vizinhos
model = KNeighborsClassifier(n_neighbors=5)
model.fit(X_train, y_train)

y_pred = model.predict(X_test)

accuracy = accuracy_score(y_test, y_pred)

precision = precision_score(y_test, y_pred)

recall = recall_score(y_test, y_pred)

f1 = f1_score(y_test, y_pred)

print('Accuracy:', accuracy)
print('Precision:', precision)
print('Recall:', recall)
print('F1 Score:', f1)

Accuracy: 0.62
Precision: 0.3333333333333333
Recall: 0.1875
F1 Score: 0.24


**Support Vector Machine**

In [14]:
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

# Cria um modelo de SVM com kernel linear
model = SVC(kernel='linear')
model.fit(X_train, y_train)

y_pred = model.predict(X_test)

accuracy = accuracy_score(y_test, y_pred)

precision = precision_score(y_test, y_pred)

recall = recall_score(y_test, y_pred)

f1 = f1_score(y_test, y_pred)

print('Accuracy:', accuracy)
print('Precision:', precision)
print('Recall:', recall)
print('F1 Score:', f1)

Accuracy: 0.66
Precision: 0.4666666666666667
Recall: 0.4375
F1 Score: 0.45161290322580644


**Decision Tree**

In [15]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

# Cria um modelo de árvore de decisão com critério de impureza de Gini e sem limite de profundidade
model = DecisionTreeClassifier(criterion='gini', max_depth=None)
model.fit(X_train, y_train)

y_pred = model.predict(X_test)

accuracy = accuracy_score(y_test, y_pred)

precision = precision_score(y_test, y_pred)

recall = recall_score(y_test, y_pred)

f1 = f1_score(y_test, y_pred)

print('Accuracy:', accuracy)
print('Precision:', precision)
print('Recall:', recall)
print('F1 Score:', f1)

Accuracy: 0.68
Precision: 0.5
Recall: 0.375
F1 Score: 0.42857142857142855


**Random Forest**

In [16]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

# Número de árvores na floresta = 100; critério de divisão = gini; profundidade máxima das árvores = ilimitada
model = RandomForestClassifier(n_estimators=100, criterion='gini', max_depth=None)
# Treina o modelo com os dados de treino
model.fit(X_train, y_train)

y_pred = model.predict(X_test)

accuracy = accuracy_score(y_test, y_pred)

precision = precision_score(y_test, y_pred)

recall = recall_score(y_test, y_pred)

f1 = f1_score(y_test, y_pred)

print('Accuracy:', accuracy)
print('Precision:', precision)
print('Recall:', recall)
print('F1 Score:', f1)

Accuracy: 0.7
Precision: 1.0
Recall: 0.0625
F1 Score: 0.11764705882352941


**MLP**

In [17]:
from sklearn.neural_network import MLPClassifier
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

# Número de neurônios na primeira camada oculta = 100; número de neurônios na segunda camada oculta = 50; função de ativação = ReLU; otimizador = Adam; número máximo de iterações = 1000
model = MLPClassifier(hidden_layer_sizes=(100, 50), activation='relu', solver='adam', max_iter=1000)
model.fit(X_train, y_train)

y_pred = model.predict(X_test)

accuracy = accuracy_score(y_test, y_pred)

precision = precision_score(y_test, y_pred)

recall = recall_score(y_test, y_pred)

f1 = f1_score(y_test, y_pred)

print('Accuracy:', accuracy)
print('Precision:', precision)
print('Recall:', recall)
print('F1 Score:', f1)

Accuracy: 0.7
Precision: 0.5714285714285714
Recall: 0.25
F1 Score: 0.34782608695652173


Após o primeiro contato, temos a opção de fazer uma **Validação Cruzada** e testar os **Hiperparâmentros**

#### Validação Cruzada

**Naive Bayes**

In [18]:
from sklearn.naive_bayes import MultinomialNB
from sklearn.model_selection import cross_val_predict
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

model = MultinomialNB()

# Faz a validação cruzada com 5 partições -> onde cada partição é usada uma vez como conjunto de teste enquanto as outras 4 são usadas como conjunto de treino
y_pred = cross_val_predict(model, X, y, cv=5)  

accuracy = accuracy_score(y, y_pred)
precision = precision_score(y, y_pred)
recall = recall_score(y, y_pred)
f1 = f1_score(y, y_pred)

print('Accuracy:', accuracy)
print('Precision:', precision)
print('Recall:', recall)
print('F1 Score:', f1)

Accuracy: 0.5080645161290323
Precision: 0.32857142857142857
Recall: 0.6216216216216216
F1 Score: 0.42990654205607476


**KNN**

In [19]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import cross_val_predict
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

# KNN com 5 vizinhos
model = KNeighborsClassifier(n_neighbors=5)  

y_pred = cross_val_predict(model, X, y, cv=5)

accuracy = accuracy_score(y, y_pred)
precision = precision_score(y, y_pred)
recall = recall_score(y, y_pred)
f1 = f1_score(y, y_pred)

print('Accuracy:', accuracy)
print('Precision:', precision)
print('Recall:', recall)
print('F1 Score:', f1)

Accuracy: 0.6209677419354839
Precision: 0.1875
Recall: 0.08108108108108109
F1 Score: 0.11320754716981132


**SVM**

In [20]:
from sklearn.svm import SVC
from sklearn.model_selection import cross_val_predict
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

# SVM com kernel linear e parâmetro de regularização C=1.0
model = SVC(kernel='linear', C=1.0)

y_pred = cross_val_predict(model, X, y, cv=5)

accuracy = accuracy_score(y, y_pred)
precision = precision_score(y, y_pred)
recall = recall_score(y, y_pred)
f1 = f1_score(y, y_pred)

print('Accuracy:', accuracy)
print('Precision:', precision)
print('Recall:', recall)
print('F1 Score:', f1)

Accuracy: 0.6290322580645161
Precision: 0.38461538461538464
Recall: 0.40540540540540543
F1 Score: 0.39473684210526316


**Decision Tree**

In [21]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import cross_val_predict
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

# Árvore de decisão com critério de impureza de Gini e sem limite de profundidade
model = DecisionTreeClassifier(criterion='gini', max_depth=None) 

y_pred = cross_val_predict(model, X, y, cv=5)

accuracy = accuracy_score(y, y_pred)
precision = precision_score(y, y_pred)
recall = recall_score(y, y_pred)
f1 = f1_score(y, y_pred)

print('Accuracy:', accuracy)
print('Precision:', precision)
print('Recall:', recall)
print('F1 Score:', f1)

Accuracy: 0.6774193548387096
Precision: 0.46153846153846156
Recall: 0.4864864864864865
F1 Score: 0.47368421052631576


**Random Forest**

In [22]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import cross_val_predict
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

# Floresta aleatória com 100 árvores, critério de impureza de Gini e sem limite de profundidade
model = RandomForestClassifier(n_estimators=100, criterion='gini', max_depth=None)

y_pred = cross_val_predict(model, X, y, cv=5)  

accuracy = accuracy_score(y, y_pred)
precision = precision_score(y, y_pred)
recall = recall_score(y, y_pred)
f1 = f1_score(y, y_pred)

print('Accuracy:', accuracy)
print('Precision:', precision)
print('Recall:', recall)


Accuracy: 0.7016129032258065
Precision: 0.5
Recall: 0.05405405405405406


**MLP**

In [23]:
from sklearn.neural_network import MLPClassifier
from sklearn.model_selection import cross_val_predict
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

# Rede neural com 100 neurônios na primeira camada oculta, 50 neurônios na segunda camada oculta, função de ativação ReLU e número máximo de iterações 1000
model = MLPClassifier(hidden_layer_sizes=(100, 50), activation='relu', max_iter=1000)  

y_pred = cross_val_predict(model, X, y, cv=5)

accuracy = accuracy_score(y, y_pred)
precision = precision_score(y, y_pred)
recall = recall_score(y, y_pred)
f1 = f1_score(y, y_pred)

print("Acurácia:", accuracy)
print("Precisão:", precision)
print("Recall:", recall)
print("F1-Score:", f1)

Acurácia: 0.5483870967741935
Precisão: 0.24324324324324326
Recall: 0.24324324324324326
F1-Score: 0.24324324324324326


#### Hiperparâmetros

Aqui é modificado os hiperparâmetros no intuito de fazer testes

**Naive Bayes**

In [24]:
# Para o modelo de Naive Bayes Multinomial, não temos possibilidades de hiperparâmetros para ajustar
from sklearn.naive_bayes import MultinomialNB
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

model = MultinomialNB()
model.fit(X_train, y_train)

y_pred = model.predict(X_test)

accuracy = accuracy_score(y_test, y_pred)

precision = precision_score(y_test, y_pred)

recall = recall_score(y_test, y_pred)

f1 = f1_score(y_test, y_pred)

print('Accuracy:', accuracy)
print('Precision:', precision)
print('Recall:', recall)
print('F1 Score:', f1)

Accuracy: 0.48
Precision: 0.35294117647058826
Recall: 0.75
F1 Score: 0.48


**KNN**

In [25]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

# Cria um modelo KNN com 15 vizinhos
model = KNeighborsClassifier(n_neighbors=15) 
model.fit(X_train, y_train)

y_pred = model.predict(X_test)

accuracy = accuracy_score(y_test, y_pred)

precision = precision_score(y_test, y_pred)

recall = recall_score(y_test, y_pred)

f1 = f1_score(y_test, y_pred)

print('Accuracy:', accuracy)
print('Precision:', precision)
print('Recall:', recall)
print('F1 Score:', f1)

Accuracy: 0.66
Precision: 0.42857142857142855
Recall: 0.1875
F1 Score: 0.2608695652173913


**SVM**

In [26]:
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

# Parâmetro de regularização C=10.0 tolerando valores na formação da margem
model = SVC(kernel='linear', C=10.0)
model.fit(X_train, y_train)

y_pred = model.predict(X_test)

accuracy = accuracy_score(y_test, y_pred)

precision = precision_score(y_test, y_pred)

recall = recall_score(y_test, y_pred)

f1 = f1_score(y_test, y_pred)

print('Accuracy:', accuracy)
print('Precision:', precision)
print('Recall:', recall)
print('F1 Score:', f1)

Accuracy: 0.66
Precision: 0.4666666666666667
Recall: 0.4375
F1 Score: 0.45161290322580644


**Decision Tree**

In [27]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

# Critério de impureza de entropia e profundidade máxima de 5
model = DecisionTreeClassifier(criterion='entropy', max_depth=5)
model.fit(X_train, y_train)

y_pred = model.predict(X_test)

accuracy = accuracy_score(y_test, y_pred)

precision = precision_score(y_test, y_pred)

recall = recall_score(y_test, y_pred)

f1 = f1_score(y_test, y_pred)

print('Accuracy:', accuracy)
print('Precision:', precision)
print('Recall:', recall)
print('F1 Score:', f1)

Accuracy: 0.56
Precision: 0.3125
Recall: 0.3125
F1 Score: 0.3125


**Random Forest**

In [28]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

# Número de árvores na floresta = 200; profundidade máxima das árvores = ilimitada
model = RandomForestClassifier(n_estimators=200, max_depth=None)
model.fit(X_train, y_train)

y_pred = model.predict(X_test)

accuracy = accuracy_score(y_test, y_pred)

precision = precision_score(y_test, y_pred)

recall = recall_score(y_test, y_pred)

f1 = f1_score(y_test, y_pred)

print('Accuracy:', accuracy)
print('Precision:', precision)
print('Recall:', recall)
print('F1 Score:', f1)

Accuracy: 0.7
Precision: 1.0
Recall: 0.0625
F1 Score: 0.11764705882352941


**MLP**

In [29]:
from sklearn.neural_network import MLPClassifier
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

# Duas camadas ocultas com 150 e 75 neurônios, função de ativação ReLU e máximo de 1000 iterações
model = MLPClassifier(hidden_layer_sizes=(150, 75), activation='relu', max_iter=1000)
model.fit(X_train, y_train)

y_pred = model.predict(X_test)

accuracy = accuracy_score(y_test, y_pred)

precision = precision_score(y_test, y_pred)

recall = recall_score(y_test, y_pred)

f1 = f1_score(y_test, y_pred)

print('Accuracy:', accuracy)
print('Precision:', precision)
print('Recall:', recall)
print('F1 Score:', f1)

Accuracy: 0.6
Precision: 0.16666666666666666
Recall: 0.0625
F1 Score: 0.09090909090909091


### Análise dos resultados

Após as operações, nessa fase vamos analisar e discutir os resultados antes do parâmetros modificados:

**Naive Bayes:**
- Acurácia: 0.48
- Precisão: 0.35
- Recall: 0.75
- F1-Score: 0.48

O Naive Bayes apresentou uma alta taxa de recall, indicando que ele é bom em identificar positivos corretos, mas a acurácia e a precisão foram relativamente baixas.

**K-Nearest Neighbors:**
- Acurácia: 0.62
- Precisão: 0.33
- Recall: 0.19
- F1-Score: 0.24

O KNN teve uma acurácia moderada, mas a precisão e o recall foram baixos, sugerindo dificuldades em lidar com falsos positivos e verdadeiros positivos.

**Support Vector Machine:**
- Acurácia: 0.66
- Precisão: 0.47
- Recall: 0.44
- F1-Score: 0.45

O SVM teve um desempenho moderado, com uma acurácia razoável, mas o recall mais baixo indica dificuldades em recuperar todos os positivos verdadeiros.

**Decision Tree:**
- Acurácia: 0.68
- Precisão: 0.50
- Recall: 0.44
- F1-Score: 0.47

A Decision Tree obteve uma acurácia intermediária, mas a precisão e o recall foram relativamente baixos, sugerindo problemas de equilíbrio entre falsos positivos e verdadeiros positivos.

**Random Forest:**
- Acurácia: 0.72
- Precisão: 1.00
- Recall: 0.13
- F1-Score: 0.22

O Random Forest teve a maior acurácia, mas a precisão muito alta e o recall muito baixo indicam um possível sobreajuste aos dados de treinamento.

**Multilayer Perceptron:**
- Acurácia: 0.66
- Precisão: 0.35
- Recall: 0.16
- F1-Score: 0.22

O Multilayer Perceptron teve uma acurácia intermediária, mas tanto a precisão quanto o recall foram baixos, sugerindo dificuldades em equilibrar falsos positivos e verdadeiros positivos.

com as mudanças dos hiperparâmetros:

**Naive Bayes:**
- Acurácia: 0.48
- Precisão: 0.35
- Recall: 0.75
- F1-Score: 0.48

Os resultados do Naive Bayes permaneceram os mesmos, pois não há hiperparâmetros para ajustar.

**K-Nearest Neighbors:**
- Acurácia: 0.60
- Precisão: 0.17
- Recall: 0.06
- F1-Score: 0.09

A mudança no número de vizinhos para 15 resultou em uma queda significativa na precisão, recall e F1-Score, indicando que o modelo teve dificuldades em identificar corretamente as classes.

**Support Vector Machine:**
- Acurácia: 0.60
- Precisão: 0.17
- Recall: 0.06
- F1-Score: 0.09

Aumentar o parâmetro de regularização C para 10.0 não melhorou o desempenho do SVM, resultando em uma baixa precisão e recall.

**Decision Tree:**
- Acurácia: 0.60
- Precisão: 0.17
- Recall: 0.06
- F1-Score: 0.09

A alteração do critério de impureza para entropia e a limitação da profundidade máxima para 5 resultaram em um desempenho pior, com baixa precisão e recall.

**Random Forest:**
- Acurácia: 0.60
- Precisão: 0.17
- Recall: 0.06
- F1-Score: 0.09

Aumentar o número de árvores para 200 não trouxe melhorias significativas, resultando em uma baixa precisão e recall.

**Multilayer Perceptron:**
- Acurácia: 0.60
- Precisão: 0.17
- Recall: 0.06
- F1-Score: 0.09

A mudança na arquitetura da rede neural para duas camadas ocultas com 150 e 75 neurônios também não melhorou o desempenho, resultando em baixa precisão e recall.
Em resumo, as mudanças nos hiperparâmetros não trouxeram melhorias significativas para os modelos, e em alguns casos, pioraram o desempenho.


Ao compararmos a complexidade do Random Forest e do MLP, podemos ter uma boa noção do desempenho dos dois. Percebemos que o conjunto de dados tinha uma boa quantidade de valores a se trabalhar para chegar à conclusão esperada. Sendo assim, a lógica do Random Forest pode ter trabalhado de maneira mais concisa ou até mais adequada frente à base de dados, enquanto a lógica do MLP pode ter se perdido na manipulação dos dados, resultando em valores baixos em recall e precisão.

Relativo ao KNN, percebemos uma acurácia moderada para baixa enquanto todos os outros indicadores tiveram um desempenho relativamente baixo. Isso pode ser justificado pela grande quantidade de características existentes na base de dados, resultando em uma distribuição desequilibrada das classes que o algoritmo usa para trabalhar.

A acurácia baixa do Naive Bayes pode ter se dado à simplicidade exacerbada do algoritmo. Isso, aliado à distribuição majoritariamente numérica do formato de dados (o que geralmente vai contra a natureza do algoritmo, que é mais utilizado para dados categóricos), pode indicar o porquê ele não conseguiu trabalhar muito bem os dados para chegar no resultado esperado.

No caso da Decision Tree, seu desempenho relativamente baixo pode ter sido dado ao fato dos conjuntos de dados terem muitos valores e não terem uma ligação direta entre si para chegar ao resultado esperado.